Resolve bug #74188 (undefined statics raising with ?? operator)

tpunt opened the pull request:
This is my attempt to fix [bug #74188](https://bugs.php.net/bug.php?id=74188). Assuming this fix is ok, I'll also submit another PR from [my other branch](https://github.com/php/php-src/compare/master...tpunt:null-coalescing-operator-with-statics) that fixes this in master, since the code in master has changed quite a lot from the 7.0 and 7.1 branches.

nikic commented on Zend/zend_vm_def.h:
```diff
@@ -1514,7 +1514,7 @@ ZEND_VM_HELPER_EX(zend_fetch_var_address_helper, CONST|TMPVAR|CV, UNUSED|CONST|V
 
 				/* check if static properties were destoyed */
-				if (UNEXPECTED(CE_STATIC_MEMBERS(ce) == NULL)) {
+				if (UNEXPECTED(CE_STATIC_MEMBERS(ce) == NULL) && type != BP_VAR_IS) {
```
retval will not be NULL in this case, it will be a dangling pointer.

nikic commented on Zend/tests/bug74188.phpt:
```diff
@@ -0,0 +1,14 @@
+--TEST--
+func_num_args() tests
```
Copy/paste mistake in test name

nikic commented on Zend/zend_vm_def.h:
```diff
@@ -1620,14 +1628,17 @@ ZEND_VM_HELPER_EX(zend_fetch_var_address_helper, CONST|TMPVAR|CV, UNUSED|CONST|V
-			ZVAL_UNREF(retval);
-		}
-		ZVAL_COPY(EX_VAR(opline->result.var), retval);
+	if (!retval) {
```
I'd prefer moving this check to only the cases where it can occur. In particular, this code is (in PHP 7.0) shared between static member fetch and non-CV variable fetches, while the NULL check is only relevant for static members. In particular, I would suggest to a) use `retval = EG(uninitialized_zval);` in the two places above where you currently have `retval = NULL;` and b) extend the NULL check at https://github.com/php/php-src/pull/2436/files#diff-3054389ad750ce9a9f5895cd6d27800fR1566 (i.e. rewrite it as `if (retval) { if (const) { cache } } else { retval = &EG(uninitialized_zval); }`).

tpunt commented:
@nikic Thanks for the comments! I have made the necessary changes now, and cleaned the history. I've also updated [my other branch targeting master](https://github.com/php/php-src/compare/master...tpunt:null-coalescing-operator-with-statics) in case it wanted.

nikic commented:
Merged as https://github.com/php/php-src/commit/744c4a5592ab3e8a733d68fbfe8913081fb6c1bc, thanks!
